src/fastalloc/vregset.rs

```diff
@@ -0,0 +1,143 @@
+use core::fmt;
+
+use crate::ion::data_structures::VRegIndex;
+use crate::VReg;
+use alloc::vec;
+use alloc::vec::Vec;
+
+#[derive(Clone)]
+struct VRegNode {
+    next: VRegIndex,
+    prev: VRegIndex,
+    vreg: VReg,
+}
+
+// Using a doubly linked list here for fast insertion,
+// removal and iteration.
+pub struct VRegSet {
+    items: Vec<VRegNode>,
+    head: VRegIndex,
+}
+
+impl VRegSet {
+    pub fn with_capacity(num_vregs: usize) -> Self {
+        Self {
+            items: vec![
+                VRegNode {
+                    prev: VRegIndex::new(num_vregs),
+                    next: VRegIndex::new(num_vregs),
+                    vreg: VReg::invalid()
+                };
+                num_vregs + 1
```

Comment:
it looks like we're using index 0 as a sentinel node (which is a great idea); do we need to add one below when we index into it with vreg numbers?

Reply:
It's the last index that's used as a sentinel, so there isn't any need to add one